api/server/httputils/errors.go

```diff
@@ -37,49 +33,45 @@ func GetHTTPErrorStatusCode(err error) int {
 	}
 
 	var statusCode int
-	errMsg := err.Error()
 
-	switch e := err.(type) {
-	case httpStatusError:
-		statusCode = e.HTTPErrorStatusCode()
-	case inputValidationError:
+	// Stop right there
+	// Are you sure you should be adding a new error class here? Do one of the existing ones work?
+
+	// Note that the below functions are already checking the error causal chain for matches.
+	switch {
```

Comment:
I understand this is a massive effort. It's great we're removing the "string contains" -> error mapping. However I don't think we should be encouraging new code to write to these error types. 

It seems like we still have this problem (from #32144):

> We end up with low-level bits worried about HTTP status codes.

A 1:1 mapping from "package error" to "http error" means we still have tight coupling between the "application logic packages" and the HTTP API. The low-level bits are still effectively communicating an HTTP error, just with a different mechanism.

I think the correct long term solution is to move away from this centralized `GetHTTPErrorStatusCode()` function entirely. We shouldn't have every error being translated in a single place.

* each package should return specific error types for each applicable error case. These error types don't need to map to anything, they just need to be distinguishable from other errors.
* the HTTP API layer (or whatever API layer) translates the specific errors for each call to the error code that is appropriate in that case.

This is important because a call could result in different error codes depending on the context in which it is used.

So I'm not sure what that means for this PR. I still think it's an improvement over string contains checking, but I think we should comment that `GetHTTPErrorStatusCode()` is deprecated and suggest the correct solution is to handle domain specific errors.

WDYT?